File: gluecodium/src/test/resources/smoke/comments/output/swift/smoke/MultiLineComments.swift
//
//
import Foundation
/// This is some very useful interface.
/// There is a lot to say about this interface.
/// at least it has multiline comments.
///
/// I am a heading
/// --------------
///
/// And now comes a list:
/// * asterisk
/// * needs
/// * escaping
///
/// ```Some example code;```
public class MultiLineComments {
    let c_instance : _baseRef
    init(cMultiLineComments: _baseRef) {
        guard cMultiLineComments != 0 else {
            fatalError("Nullptr value is not supported for initializers")
        }
        c_instance = cMultiLineComments
    }
    deinit {
        smoke_MultiLineComments_remove_swift_object_from_wrapper_cache(c_instance)
        smoke_MultiLineComments_release_handle(c_instance)
    }
    /// This is very important method.
    /// It has very important parameters.
    /// It has side effects.
    /// - Parameters:
    ///   - input: Very useful input parameter.
    ///       You must not confuse it with the second parameter.
    ///       But they are similar.
    ///   - ratio: Not as useful as the first parameter.
    ///       But still useful.
    ///       use a positive value for more happiness.
    /// - Returns: If you provide a useful input,
    ///     and a useful ratio you can expect a useful output.
    ///     Just kidding do not expect anything from a method until
    ///     you see its body.
    public func someMethodWithLongComment(input: String, ratio: Double) -> Float {
        let c_input = moveToCType(input)
        let c_ratio = moveToCType(ratio)
        return moveFromCType(smoke_MultiLineComments_someMethodWithLongComment(self.c_instance, c_input.ref, c_ratio.ref))
    }
}
internal func getRef(_ ref: MultiLineComments?, owning: Bool = true) -> RefHolder {
    guard let c_handle = ref?.c_instance else {
        return RefHolder(0)
    }
    let handle_copy = smoke_MultiLineComments_copy_handle(c_handle)
    return owning
        ? RefHolder(ref: handle_copy, release: smoke_MultiLineComments_release_handle)
        : RefHolder(handle_copy)
}
extension MultiLineComments: NativeBase {
    var c_handle: _baseRef { return c_instance }
}
extension MultiLineComments: Hashable {
    // :nodoc:
    public static func == (lhs: MultiLineComments, rhs: MultiLineComments) -> Bool {
        return lhs.c_handle == rhs.c_handle
    }
    // :nodoc:
    public func hash(into hasher: inout Hasher) {
        hasher.combine(c_handle)
    }
}
internal func foobar_MultiLineComments_copyFromCType(_ handle: _baseRef) -> MultiLineComments {
    if let swift_pointer = smoke_MultiLineComments_get_swift_object_from_wrapper_cache(handle),
        let re_constructed = Unmanaged<AnyObject>.fromOpaque(swift_pointer).takeUnretainedValue() as? MultiLineComments {
        return re_constructed
    }
    let result = MultiLineComments(cMultiLineComments: smoke_MultiLineComments_copy_handle(handle))
    smoke_MultiLineComments_cache_swift_object_wrapper(handle, Unmanaged<AnyObject>.passUnretained(result).toOpaque())
    return result
}
internal func foobar_MultiLineComments_moveFromCType(_ handle: _baseRef) -> MultiLineComments {
    if let swift_pointer = smoke_MultiLineComments_get_swift_object_from_wrapper_cache(handle),
        let re_constructed = Unmanaged<AnyObject>.fromOpaque(swift_pointer).takeUnretainedValue() as? MultiLineComments {
        smoke_MultiLineComments_release_handle(handle)
        return re_constructed
    }
    let result = MultiLineComments(cMultiLineComments: handle)
    smoke_MultiLineComments_cache_swift_object_wrapper(handle, Unmanaged<AnyObject>.passUnretained(result).toOpaque())
    return result
}
internal func foobar_MultiLineComments_copyFromCType(_ handle: _baseRef) -> MultiLineComments? {
    guard handle != 0 else {
        return nil
    }
    return foobar_MultiLineComments_moveFromCType(handle) as MultiLineComments
}
internal func foobar_MultiLineComments_moveFromCType(_ handle: _baseRef) -> MultiLineComments? {
    guard handle != 0 else {
        return nil
    }
    return foobar_MultiLineComments_moveFromCType(handle) as MultiLineComments
}
internal func foobar_copyToCType(_ swiftClass: MultiLineComments) -> RefHolder {
    return getRef(swiftClass, owning: false)
}
internal func foobar_moveToCType(_ swiftClass: MultiLineComments) -> RefHolder {
    return getRef(swiftClass, owning: true)
}
internal func foobar_copyToCType(_ swiftClass: MultiLineComments?) -> RefHolder {
    return getRef(swiftClass, owning: false)
}
internal func foobar_moveToCType(_ swiftClass: MultiLineComments?) -> RefHolder {
    return getRef(swiftClass, owning: true)
}

File: gluecodium/src/test/resources/smoke/comments/output/dart/lib/src/smoke/multi_line_comments.dart
import 'dart:ffi';
import 'package:library/src/_library_context.dart' as __lib;
import 'package:library/src/_native_base.dart' as __lib;
import 'package:library/src/_token_cache.dart' as __lib;
import 'package:library/src/builtin_types__conversion.dart';
/// This is some very useful interface.
///
/// There is a lot to say about this interface.
/// at least it has multiline comments.
///
/// I am a heading
/// --------------
///
/// And now comes a list:
/// * asterisk
/// * needs
/// * escaping
///
/// ```Some example code;```
abstract class MultiLineComments {

  /// This is very important method.
  ///
  /// It has very important parameters.
  /// It has side effects.
  ///
  /// [input] Very useful input parameter.
  /// You must not confuse it with the second parameter.
  /// But they are similar.
  ///
  /// [ratio] Not as useful as the first parameter.
  /// But still useful.
  /// use a positive value for more happiness.
  ///
  /// Returns [double]. If you provide a useful input,
  /// and a useful ratio you can expect a useful output.
  /// Just kidding do not expect anything from a method until
  /// you see its body.
  ///
  double someMethodWithLongComment(String input, double ratio);
}
// MultiLineComments "private" section, not exported.
final _smokeMultilinecommentsRegisterFinalizer = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>, Int32, Handle),
    void Function(Pointer<Void>, int, Object)
  >('library_smoke_MultiLineComments_register_finalizer'));
final _smokeMultilinecommentsCopyHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_MultiLineComments_copy_handle'));
final _smokeMultilinecommentsReleaseHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_MultiLineComments_release_handle'));
class MultiLineComments$Impl extends __lib.NativeBase implements MultiLineComments {
  MultiLineComments$Impl(Pointer<Void> handle) : super(handle);

  @override
  double someMethodWithLongComment(String input, double ratio) {
    final _someMethodWithLongCommentFfi = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<Float Function(Pointer<Void>, Int32, Pointer<Void>, Double), double Function(Pointer<Void>, int, Pointer<Void>, double)>('library_smoke_MultiLineComments_someMethodWithLongComment__String_Double'));
    final _inputHandle = stringToFfi(input);
    final _ratioHandle = (ratio);
    final _handle = this.handle;
    final __resultHandle = _someMethodWithLongCommentFfi(_handle, __lib.LibraryContext.isolateId, _inputHandle, _ratioHandle);
    stringReleaseFfiHandle(_inputHandle);
    try {
      return (__resultHandle);
    } finally {
    }
  }
}
Pointer<Void> smokeMultilinecommentsToFfi(MultiLineComments value) =>
  _smokeMultilinecommentsCopyHandle((value as __lib.NativeBase).handle);
MultiLineComments smokeMultilinecommentsFromFfi(Pointer<Void> handle) {
  final instance = __lib.getCachedInstance(handle);
  if (instance != null && instance is MultiLineComments) return instance;
  final _copiedHandle = _smokeMultilinecommentsCopyHandle(handle);
  final result = MultiLineComments$Impl(_copiedHandle);
  __lib.cacheInstance(_copiedHandle, result);
  _smokeMultilinecommentsRegisterFinalizer(_copiedHandle, __lib.LibraryContext.isolateId, result);
  return result;
}
void smokeMultilinecommentsReleaseFfiHandle(Pointer<Void> handle) =>
  _smokeMultilinecommentsReleaseHandle(handle);
Pointer<Void> smokeMultilinecommentsToFfiNullable(MultiLineComments? value) =>
  value != null ? smokeMultilinecommentsToFfi(value) : Pointer<Void>.fromAddress(0);
MultiLineComments? smokeMultilinecommentsFromFfiNullable(Pointer<Void> handle) =>
  handle.address != 0 ? smokeMultilinecommentsFromFfi(handle) : null;
void smokeMultilinecommentsReleaseFfiHandleNullable(Pointer<Void> handle) =>
  _smokeMultilinecommentsReleaseHandle(handle);
// End of MultiLineComments "private" section.
